theano/sandbox/cuda/cusolver.py

```diff
@@ -0,0 +1,92 @@
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[])()]
+
+    def make_node(self, inp):
+        inp = as_cuda_ndarray_variable(inp)
+        assert inp.ndim == 2
+
+        return theano.Apply(self, [inp], self.output_type(inp))
+
+    def make_thunk(self, node, storage_map, _, no_recycling=[]):
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        def thunk():
+            A = inputs[0][0]
+            m, n = A.shape
+            # LAPACK routines expect Fortran order
+            A_T = dimshuffle(A, (1, 0)).reshape((m, n))
+            A_copy = A_T.copy()
+
+            lda = max(1, m)
+            TAU = CudaNdarray.zeros((min(m, n),))
+            handle = cusolverDnCreate()
+            Lwork = cusolverDnSgeqrf_bufferSize(handle, m, n, A_copy.gpudata,
+                                                lda)
+            Work = CudaNdarray.zeros((max(1, Lwork),))
+            devInfo = CudaNdarray.zeros((1,))
+            cusolverDnSgeqrf(handle, m, n, A_copy.gpudata, lda, TAU.gpudata,
+                             Work.gpudata, Lwork, devInfo.gpudata)
+            cusolverDnDestroy(handle)
+            outputs[0][0] = reshape(A_copy, (n, m)).dimshuffle((1, 0))
+            outputs[1][0] = TAU
+            outputs[2][0] = Work
+            outputs[3][0] = devInfo
+
+        return thunk
+
+
+geqrf = GpuGeqrf()
+
+
+_libcusolver.cusolverDnSgeqrf.restype = int
+_libcusolver.cusolverDnSgeqrf.argtypes = [ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p,
+                                          ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p]
+
+
+def cusolverDnSgeqrf(handle, m, n, A, lda, TAU, Work, Lwork, devInfo):
+    status = _libcusolver.cusolverDnSgeqrf(handle, m, n, int(A), lda,
+                                           int(TAU), int(Work),
+                                           Lwork, int(devInfo))
+    cusolverCheckStatus(status)
+
+
+_libcusolver.cusolverDnSgeqrf_bufferSize.restype = int
+_libcusolver.cusolverDnSgeqrf_bufferSize.argtypes = [ctypes.c_void_p,
+                                                     ctypes.c_int,
+                                                     ctypes.c_int,
+                                                     ctypes.c_void_p,
+                                                     ctypes.c_int,
+                                                     ctypes.c_void_p]
+
+
+def cusolverDnSgeqrf_bufferSize(handle, m, n, A, lda):
```

Comment:
As this function is very small and currently used in only one place, why you didn't inline it in the thunk function?